drivers/flash/flash_sam0.c

```diff
@@ -244,25 +236,30 @@ static int flash_sam0_write(const struct device *dev, off_t offset,
 		return err;
 	}
 
+	if (len == 0) {
+		return 0;
+	}
+
```

Comment:
The `len == 0` check should not be required and gracefully handled by the writing routine. These checks are polluting the code and are a bad idea:
1. Trying to return quickly on a condition that should not appear is adding a comparison to all real writes,
2. Handling the write of `len == 0` as a normal write could enable detection that there is something wrong with the interface without doing a real write.

Reply:
I'd suggest that this _is_ the "_graceful_" handling of a zero-length write...

> Trying to return quickly on a condition that should not appear is adding a comparison to all real writes,

Its impact is incredibly minimal... but if others agree, I'll remove it.
I think the sensible alternative to an early return is for **_two_** guards - one around each `flash_sam0_write_protection()`.

> Handling the write of `len == 0` as a normal write could enable detection that there is something wrong with the interface without doing a real write.

You can't get much closer to the lower level calls than this - `flash_sam0_commit()` won't be called with `len == 0` anyway... and as above, it would be sensible to add more guards.

Reply:
@attie-argentum, what I mean by graceful handling is that whatever follows is not influenced by this check (and I think it is already handled, nothing will happen if the line is removed). There shouldn't be an "early return", why should there be, it is no error condition.

An extra guard is also not required, if the flash_sam0_write_protection would fail for a write of 0 this is interesting knowledge that could be usefull. It would allow using a write of zero length to verify if it can write. The same holds for trying to write outside of the region, it will return an error message.

Reply:
I'm not going to argue the point too hard - I'll remove the check if needs be...

I suppose I prefer to protect the hardware from bad users - i.e: the unnecessary "_unprotect_" &rarr; "_re-protect_" process that is incurred by a zero-length write, which would be detrimental to the device's lifespan if it's over-used...

If you're trying to write, why not just try to write? A zero-length write only checks that the start is in-range, not the whole target region, so it's of limited use (but `flash_sam0_valid_range()` is still called in this case).

The docs don't mention anything about zero-length writes ([link](https://docs.zephyrproject.org/3.2.0/hardware/peripherals/flash.html#c.flash_write)), and in this mode (with `SOC_FLASH_SAM0_EMULATE_BYTE_PAGES` defined), we're already doing a lot of hand-holding for the user...

In fact, other drives implement this check too! 🙃

https://github.com/zephyrproject-rtos/zephyr/blob/e5efa3dfffd8d3eba0ae4bb580c535dadacfc7f9/drivers/flash/flash_sam.c#L180-L182

https://github.com/zephyrproject-rtos/zephyr/blob/e5efa3dfffd8d3eba0ae4bb580c535dadacfc7f9/drivers/flash/flash_stm32.c#L220-L222

https://github.com/zephyrproject-rtos/zephyr/blob/e5efa3dfffd8d3eba0ae4bb580c535dadacfc7f9/drivers/flash/flash_gd32.c#L58-L60

Reply:
> I'm not going to argue the point too hard - I'll remove the check if needs be...
> 

Neither am I. 😀 

> I suppose I prefer to protect the hardware from bad users - i.e: the unnecessary "_unprotect_" → "_re-protect_" process that is incurred by a zero-length write, which would be detrimental to the device's lifespan if it's over-used...
>

That is a good reason, and should be the only reason to do this.
 
> In fact, other drives implement this check too! 🙃

I know, but this shouldn't stop you from trying to do better. 😀

Reply:
> but this shouldn't stop you from trying to do better

Now there's a controversial statement... I thought I was trying to do better 😜

... I'll let others chip in on the `len == 0` test, I hope that's okay.